lib/gc/Transforms/Microkernel/ConvertLinalgToMicrokernel.cpp

```diff
@@ -0,0 +1,390 @@
+}
+
+// Return the position of `dim` in the codomain of `operand`.
+static std::optional<unsigned> getPosInCodomain(ArrayRef<unsigned> dimPos,
+                                                OpOperand *operand,
+                                                linalg::LinalgOp linalgOp) {
+  assert(operand->getOwner() == linalgOp);
+  auto map = linalgOp.getMatchingIndexingMap(operand);
+  for (unsigned i = 0, numResults = map.getNumResults(); i < numResults; i++) {
+    if (isMatchingAffineResult(linalgOp, map.getResult(i), dimPos))
+      return i;
+  }
+  return std::nullopt;
+}
+
+static FailureOr<BrgemmInfo>
+inferBrgemmInfo(linalg::LinalgOp linalgOp,
+                const linalg::ContractionDimensions &dims) {
+  unsigned mPos = dims.m[0];
+  unsigned nPos = dims.n[0];
+  // dims.k could be of 2 cases:
+  //     1. dims.k.size() == 2: non-VNNI, K = dims.k[1]
+  //     2. dims.k.size() == 3: VNNI, K = dims.k[1] * dims.k[2]
+  unsigned batchPos = dims.k.front();
+  SmallVector<unsigned, 2> kPos;
+  if (dims.k.size() == 2) {
+    kPos = {dims.k[1]};
+  } else if (dims.k.size() == 3) {
+    kPos = {dims.k[1], dims.k[2]};
+  } else {
+    return failure();
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmInfo] Candidate dims: "
+                          << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmInfo] m pos in affine: " << mPos
+                          << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmInfo] n pos in affine: " << nPos
+                          << "\n");
+  for (auto kp : kPos) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "[inferBrgemmInfo] k pos in affine: " << kp << "\n");
+  }
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmInfo] batch pos in affine: "
+                          << batchPos << "\n");
+
+  auto checkStridesAndGetLda =
+      [&](ArrayRef<unsigned> minorDim, ArrayRef<unsigned> majorDim,
+          OpOperand *operand, bool allowVnni) -> FailureOr<int64_t> {
+    auto minorDimPosInCodomain = getPosInCodomain(minorDim, operand, linalgOp);
+    auto majorDimPosInCodomain = getPosInCodomain(majorDim, operand, linalgOp);
+    if (!minorDimPosInCodomain || !majorDimPosInCodomain)
+      return failure();
+    auto stridesOnOperand = utils::getStaticStrides(operand->get());
+    if (failed(stridesOnOperand))
+      return failure();
+    auto minorDimLd = (*stridesOnOperand)[*minorDimPosInCodomain];
+    auto majorDimLd = (*stridesOnOperand)[*majorDimPosInCodomain];
+    if (minorDimLd != 1) {
+      // VNNI format exists, special treatment to align LD with non-VNNI format
+      if (!allowVnni || (minorDimLd != 2 && minorDimLd != 4))
+        return failure();
+      return majorDimLd / minorDimLd;
+    }
+    return majorDimLd;
+  };
+
+  OpOperand *operandA = linalgOp.getDpsInputOperands()[0];
+  OpOperand *operandB = linalgOp.getDpsInputOperands()[1];
+  OpOperand *operandC = &linalgOp.getDpsInitsMutable()[0];
+
+  // A(m, k)
+  auto lda = checkStridesAndGetLda(kPos, {mPos}, operandA, false);
```

Comment:
Could you please put a real type whenever it is not obvious? (it's obvious when you can see the type on the right side, e.g., a create<some_type>). Other places too, not marking all of them.